Library/ViewModels/DashboardViewModel.swift

```diff
@@ -87,25 +102,65 @@ public protocol DashboardViewModelType {
 public final class DashboardViewModel: DashboardViewModelInputs, DashboardViewModelOutputs,
   DashboardViewModelType {
 
-    public init() {
+  //We use producers ot this properties during view appearance and projects retreaval process 
+  //to avoid any race condisions and keep last selected project param "handy" 
+  //for the moment projects are reteaved
+  private let selectedProjectProperty: SignalProducer<Param?, NoError>
+  private let selectProjectPropertyOrFirst = MutableProperty<Param?>(nil)
+  private let messageThreadReceived = MutableProperty<(Param, MessageThread)?>(nil)
+
+  public init() {
     let projects = self.viewWillAppearAnimatedProperty.signal.filter(isFalse).ignoreValues()
       .switchMap {
         AppEnvironment.current.apiService.fetchProjects(member: true)
           .ksr_delay(AppEnvironment.current.apiDelayInterval, on: AppEnvironment.current.scheduler)
           .demoteErrors()
           .map { $0.projects }
+          .prefix(value: [])
     }
 
+    self.selectedProjectProperty = SignalProducer.merge(
+      self.switchToProjectProperty.producer,
+      self.messageThreadNavigatedProperty.producer.skipNil().map { $0.0 }
+    )
+
+    self.selectProjectPropertyOrFirst <~ SignalProducer.combineLatest(
+      self.selectedProjectProperty,
+      self.viewWillAppearAnimatedProperty.producer.ignoreValues()
+    )
+    .map { $0.0 }
+    .skipRepeats { lhs, rhs in lhs == rhs }
+
     let projectsAndSelected = projects
-      .switchMap { [switchToProject = self.switchToProjectProperty.producer] projects in
+      .switchMap { [switchToProject = self.selectProjectPropertyOrFirst.producer] projects in
         switchToProject
-          .map { param in
-            find(projectForParam: param, in: projects) ?? projects.first
+          .map { param -> Project? in
+            guard let param = param else {
```

Comment:
Can use optional composition here over `guard let` and avoid the extra `projects.first` return:

``` swift
param.flatMap { find(projectForParam: $0, in: projects) } ?? projects.first
```